syntax = "proto3";
package gladys.v1;

import "common.proto";
import "types.proto";

// ExecutiveService - the decision-making component.
// Receives events from Orchestrator and processes them.
service ExecutiveService {
    // Process an event (queued by EventQueue, processed by salience priority)
    rpc ProcessEvent(ProcessEventRequest) returns (ProcessEventResponse);

    // Provide feedback on a previous response (for heuristic formation)
    rpc ProvideFeedback(ProvideFeedbackRequest) returns (ProvideFeedbackResponse);

    // Health check
    rpc GetHealth(gladys.types.GetHealthRequest) returns (gladys.types.GetHealthResponse);
    rpc GetHealthDetails(gladys.types.GetHealthDetailsRequest) returns (gladys.types.GetHealthDetailsResponse);
}

// Suggestion from a low-confidence heuristic match
// Passed to Executive so LLM can consider the pattern
message HeuristicSuggestion {
    string heuristic_id = 1;        // UUID of the matched heuristic
    string suggested_action = 2;    // From effects_json.message
    float confidence = 3;           // Current confidence (< threshold)
    string condition_text = 4;      // What pattern matched
}

// Request to process a single high-salience event
message ProcessEventRequest {
    Event event = 1;
    bool immediate = 2;
    HeuristicSuggestion suggestion = 3;  // Optional: low-conf heuristic that matched
    repeated HeuristicSuggestion candidates = 4;  // Other below-threshold heuristic matches
    RequestMetadata metadata = 15;
}

message ProcessEventResponse {
    bool accepted = 1;
    string error_message = 2;
    string response_id = 3;
    string response_text = 4;
    float predicted_success = 5;
    float prediction_confidence = 6;
    string prompt_text = 7;           // Full LLM prompt (empty for heuristic fast-path)
    string decision_path = 8;         // "heuristic" or "llm"
    string matched_heuristic_id = 9;  // UUID of involved heuristic (empty if none)
}

// Request to provide feedback on a previous LLM response
message ProvideFeedbackRequest {
    string event_id = 1;
    bool positive = 2;
    string response_id = 3;
    RequestMetadata metadata = 15;
}

message ProvideFeedbackResponse {
    bool accepted = 1;
    string error_message = 2;
    string created_heuristic_id = 3;
}
